server/model/maintenance.js

```diff
@@ -253,6 +253,10 @@ class Maintenance extends BeanModel {
                         duration = this.duration * 1000;
                     }
 
+                    if (duration === undefined && this.strategy === "recurring-interval") {
```

Comment:
I need a bit of help understanding the `if`-clause you added 😅:
Is this not already covered by the `if` or `else` branch above?

Reply:
Hi, @CommanderStorm, so above is an `if` - `else if` - `else`, and the `else if` is when `this.end_date` is set, and it calculates if `this.end_date` will be before the end of the interval. If it is not, then `this.duration` is left `undefined`.
I guess my solution is not the best one, I was not sure if it was left `undefined` for some other purpose, but if it is not, then I guess the `else if` should look something like this:

```javascript
else if (this.end_date) {
    let d = dayjs(this.end_date).diff(dayjs(), "second");
    if (d < this.duration) {
        duration = d * 1000;
    }
    else {
        duration = this.duration * 1000;
    }
}
```
Let me know if you want me to make this change.

Reply:
I think that is better.
`duration` is never explicitly checked for undefined => I think this is a bug ^^

While you are at it, I think that while `if`/... clause would be way better if extracted into a function `this.duration = this.inferDuration()`
That way the controll flow is not as confusing in the future..

Reply:
```javascript
    inferDuration(customDuration) {
        // Check if duration is still in the window. If not, use the duration from the current time to the end of the window
        if (customDuration > 0) {
            return customDuration;
        } else if (this.end_date) {
            let d = dayjs(this.end_date).diff(dayjs(), "second");
            if (d < this.duration) {
                return d * 1000;
            }
        }
        return this.duration * 1000;
    }
````
[Code here](https://github.com/buzzinJohnnyBoi/uptime-kuma/blob/30f7b4898758ea5422983d00a3d3a3597e7e1524/server/model/maintenance.js#L319C5-L330C6)\
Lmk if this is good.